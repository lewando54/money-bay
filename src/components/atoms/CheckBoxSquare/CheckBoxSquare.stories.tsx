/* eslint-disable @typescript-eslint/consistent-type-assertions */
import React, { useState } from 'react'
import { Meta,StoryObj } from '@storybook/react-native'

import CheckBoxSquare from './CheckBoxSquare'

import { GOOD_GREEN } from '@styling/GlobalTheme.style'

interface ICheckBoxSquareMockupProps {
    checked?: boolean
    borderColor?: string
    checkedColor?: string
    checkmarkColor?: string
    onPress?: () => void
}

function StatefulChceckbox({
    checked=false,
    borderColor='#868698',
    checkedColor=GOOD_GREEN,
    checkmarkColor='white',
    onPress,
}: ICheckBoxSquareMockupProps){
    const [checkedState, setCheckedState] = useState(checked)

    return (
        <CheckBoxSquare 
            checked={checkedState} 
            borderColor={borderColor}
            checkedColor={checkedColor}
            checkmarkColor={checkmarkColor}
            onPress={(newValue) => {onPress(); setCheckedState(() => newValue)}}
        />
    )
}

const meta = {
    component: StatefulChceckbox,
    argTypes: {
        checked: { control: 'boolean' },
        borderColor: { control: 'color' },
        checkedColor: { control: 'color' },
        checkmarkColor: { control: 'color' },
        onPress: { action: 'pressed' }
    }
} satisfies Meta<typeof StatefulChceckbox>

export default meta
type Story = StoryObj<typeof StatefulChceckbox>

export const Default: Story = {
    args: {
        checked: false,
        borderColor: '#868698',
        checkedColor: GOOD_GREEN,
        checkmarkColor: 'white',
    }
}

export const CustomColor: Story = {
    args: {
        checked: false,
        borderColor: 'magenta',
        checkedColor: 'black',
        checkmarkColor: 'red',
    }
}
