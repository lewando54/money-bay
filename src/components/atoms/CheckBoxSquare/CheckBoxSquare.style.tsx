import { StyleSheet } from 'react-native'
export default StyleSheet.create({
    container: {
        width: 22,
        height: 22,
        borderWidth: 1,
        borderRadius: 4
    }
})
