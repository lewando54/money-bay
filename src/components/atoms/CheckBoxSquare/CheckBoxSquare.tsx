import React from 'react'
import { Pressable } from 'react-native'
import Animated, { Easing, interpolate, interpolateColor, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'
import { Feather } from '@expo/vector-icons'

import { GOOD_GREEN } from '../../../styling/GlobalTheme.style'
import checkBoxSquareStyle from './CheckBoxSquare.style'

interface ICheckBoxSquareProps {
    checked?: boolean
    borderColor?: string
    checkedColor?: string
    checkmarkColor?: string
    onPress?: (newValue: boolean) => void
    testID?: string
}

export default function CheckBoxSquare({
    checked=false,
    borderColor='#868698',
    checkedColor=GOOD_GREEN,
    checkmarkColor='white',
    onPress,
    testID
}: ICheckBoxSquareProps){
    const animationProgress = useSharedValue(0)

    const checkmarkAnimationStyle = useAnimatedStyle(() => ({
        transform: [{ rotateZ: interpolate(animationProgress.value, [0, 1], [180, 0]) + 'deg' }],
        opacity: animationProgress.value
    }))

    const animatedBackgroundColorStyle = useAnimatedStyle(() => ({
        backgroundColor: interpolateColor(
            animationProgress.value,
            [0, 1],
            ['transparent', checkedColor]
        ),
        borderColor: interpolateColor(
            animationProgress.value,
            [0, 1],
            [borderColor, checkedColor]
        )
    }))

    const handleOnPress = () => {
        if(onPress) onPress(!checked)

        animationProgress.value = withTiming(1 - animationProgress.value, {duration: 500, easing: Easing.inOut(Easing.exp)})
    }

    return (
        <Pressable onPress={handleOnPress} testID={testID}>
            <Animated.View style={[checkBoxSquareStyle.container, {borderColor: borderColor, justifyContent: 'center', alignItems: 'center'}, animatedBackgroundColorStyle]}>
                <Animated.View style={[checkmarkAnimationStyle]}>
                    <Feather name="check" size={18} color={checkmarkColor} />
                </Animated.View>
            </Animated.View>
        </Pressable>
    )
}
