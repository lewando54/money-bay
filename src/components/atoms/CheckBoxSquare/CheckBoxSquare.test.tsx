import React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react-native'
import CheckBoxSquare from './CheckBoxSquare'
describe('CheckBoxSquare', () => {
    jest.useFakeTimers()
    it('should render correctly', () => {
        const tree = renderer.create(<CheckBoxSquare />)
        expect(tree).toMatchSnapshot()
    })

    it('should be checked when pressed', () => {
        const setChecked = jest.fn()
        const { getByTestId } = render(<CheckBoxSquare checked={false} onPress={(newValue) => setChecked(newValue)} testID='test'/>)
        fireEvent.press(getByTestId('test'))
        expect(setChecked).toHaveBeenCalledWith(true)
    })

    it('should be unchecked when pressed while checked', () => {
        const setChecked = jest.fn()
        const { getByTestId } = render(<CheckBoxSquare checked={true} onPress={(newValue) => setChecked(newValue)} testID='test'/>)
        fireEvent.press(getByTestId('test'))
        expect(setChecked).toHaveBeenCalledWith(false)
    })
})
