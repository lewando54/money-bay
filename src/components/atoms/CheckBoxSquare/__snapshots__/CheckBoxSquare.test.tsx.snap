// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxSquare should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    animatedStyle={
      {
        "value": {
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "borderColor": "rgba(134, 134, 152, 1)",
        },
      }
    }
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "borderColor": "rgba(134, 134, 152, 1)",
        "borderRadius": 4,
        "borderWidth": 1,
        "height": 22,
        "justifyContent": "center",
        "width": 22,
      }
    }
  >
    <View
      animatedStyle={
        {
          "value": {
            "opacity": 0,
            "transform": [
              {
                "rotateZ": "180deg",
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        {
          "opacity": 0,
          "transform": [
            {
              "rotateZ": "180deg",
            },
          ],
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
